File: src/DataScience/MATLAB/interpreter.m
function result = interpreter
%INTERPRETER Evaluate a tiny arithmetic language represented as an AST.
expression = binary("add", numberNode(7), binary("multiply", numberNode(3), numberNode(4)));
result = struct("value", evaluate(expression), "expression", "7+(3*4)");
end

function node = numberNode(value)
node = struct("kind", "number", "value", value, "left", [], "right", []);
end

function node = binary(kind, left, right)
node = struct("kind", kind, "value", 0, "left", left, "right", right);
end

function value = evaluate(node)
switch node.kind
    case "number"
        value = node.value;
    case "add"
        value = evaluate(node.left) + evaluate(node.right);
    case "multiply"
        value = evaluate(node.left) * evaluate(node.right);
    otherwise
        error("Unknown expression kind: %s", node.kind);
end
end

File: src/DataScience/Julia/interpreter.jl
abstract type Expr end

struct Number <: Expr
    value::Int
end

struct Add <: Expr
    left::Expr
    right::Expr
end

interpret(expr::Number) = expr.value
interpret(expr::Add) = interpret(expr.left) + interpret(expr.right)

expr = Add(Number(2), Add(Number(3), Number(4)))
value = interpret(expr)
value == 9 || error("expected 9, got $value")
println("value=$value")
